onnxscript/function_libs/torch_aten/ops/core.py

```diff
@@ -3572,17 +3573,180 @@ def aten_max_pool1d_with_indices(
+        strides = [stride] * expand_size
+    elif stride is None:
+        strides = kernel_shape
+    else:
+        strides = stride
+
+    return _aten_max_pool2d_onnx(self, kernel_shape, strides, pads, dilations, ceil_mode)
+
+
+@torch_op("aten::max_pool2d", private=True)
+def _aten_max_pool2d_onnx(
+    self: TFloatOrUInt8,
+    kernel_shape: Sequence[int],
+    strides: Sequence[int],
+    pads: Sequence[int],
+    dilations: Sequence[int],
+    ceil_mode: bool,
+) -> TFloatOrUInt8:
+    self_rank = op.Size(op.Shape(self))
+    if self_rank == 3:  # C,H,W -> N,C,H,W and N=1
+        self = op.Unsqueeze(self, op.Constant(value_ints=[0]))
+
+    pool_result, _ = op.MaxPool(
+        self,
+        ceil_mode=ceil_mode,
+        dilations=dilations,
+        kernel_shape=kernel_shape,
+        pads=pads,
+        strides=strides,
+    )
+
+    if self_rank == 3:
+        pool_result = op.Squeeze(pool_result, op.Constant(value_ints=[0]))
+
+    return pool_result
+
+
+@torch_op("aten::max_pool2d", trace_only=True)
+def aten_max_pool2d_with_indices(
+    self: TFloatOrUInt8,
+    kernel_size: Sequence[int],
+    stride: Sequence[int] = (),
+    padding: Sequence[int] = (0, 0),
+    dilation: Sequence[int] = (1, 1),
+    ceil_mode: bool = False,
+) -> tuple[TFloatOrUInt8, INT64]:
+    """max_pool2d(Tensor self, int[2] kernel_size, int[2] stride=[], int[2] padding=0, int[2] dilation=1, bool ceil_mode=False) -> (Tensor, Tensor)"""
+
+    # Torch prefer to use single number x for kerne,stride,pad,dilation on both side implicitly
+    # But ONNX needs pair number [x,y] to specify on each side explicitly
+    # For pool3d, this number should be 3
+    expand_size = 2
+
+    # The dilations should be [x, y]
+    if isinstance(dilation, int):  # x -> [x, x]
+        dilations = [dilation] * expand_size
+    else:  # already [x, y]
+        dilations = dilation
+
+    # The kernel_shape should be [x, y]
+    if isinstance(kernel_size, int):  # x -> [x, x]
+        kernel_shape = [kernel_size] * expand_size
+    else:  # assert(len(kernel_size)==2), already [x, y]
+        kernel_shape = kernel_size
+
+    # The pads should be [w, x, y, z]
+    if isinstance(padding, int):  # w -> [w, w, w, w]
+        pads = [padding] * expand_size * 2
+    elif len(padding) == 1:  # [w] -> [w, w, w, w]
+        pads = padding * 4
+    elif len(padding) == 2:  # [w, x] -> [w, x, w, x]
+        pads = padding * 2
+    else:  # assert len(padding) == 4, already [w, x, y, z]
+        pads = padding
+
+    # The strides should be [x, y]
+    if isinstance(stride, int):  # x -> [x, x]
+        strides = [stride] * expand_size
+    elif stride is None:
+        strides = kernel_shape
+    else:
+        strides = stride
+
+    return _aten_max_pool2d_with_indices_onnx(
+        self, expand_size, kernel_shape, strides, pads, dilations, ceil_mode
+    )
+
+
+@torch_op("aten::max_pool2d", private=True)
+def _aten_max_pool2d_with_indices_onnx(
```

Comment:
This function is breaking CI @justinchuby @xiaowuhu. It seems like the ONNXFunction is invalid. Maybe `if self.rank == 3` makes this function trace_only?

Reply:
Good catch. So instead of making it trace_only, we should find ways to script the function. In this case rank = op.Size(op.Shape())